test/hooks.test.js

```diff
@@ -259,6 +259,98 @@ test('preHandler hook should support encapsulation / 5', t => {
   })
 })
 
+test('onRoute hook should be called / 1', t => {
+  t.plan(2)
+  const fastify = Fastify()
+
+  fastify.register((instance, opts, next) => {
+    instance.addHook('onRoute', () => {
+      t.pass()
+    })
+    instance.get('/', opts, function (req, reply) {
+      reply.send()
+    })
+    next()
+  })
+
+  fastify.ready(err => {
+    t.error(err)
+  })
+})
+
+test('onRoute hook should be called / 2', t => {
+  t.plan(3)
+  const fastify = Fastify()
+  fastify.addHook('onRoute', (route) => {
+    t.pass()
+  })
+
+  fastify.register((instance, opts, next) => {
+    instance.addHook('onRoute', (route) => {
+      t.pass()
+    })
+    instance.get('/', opts, function (req, reply) {
+      reply.send()
+    })
+    next()
+  })
+
+  fastify.ready(err => {
+    t.error(err)
+  })
+})
+
+test('onRoute should keep the context', t => {
+  t.plan(4)
+  const fastify = Fastify()
+  fastify.register(plugin)
+
+  function plugin (instance, opts, next) {
+    instance.decorate('test', true)
+    instance.addHook('onRoute', onRoute)
+    t.ok(instance.prototype === fastify.prototype)
+
+    function onRoute (route) {
+      t.ok(this.test)
+      t.strictEqual(this, instance)
+    }
+
+    instance.get('/', opts, function (req, reply) {
+      reply.send()
+    })
+
+    next()
+  }
+
+  fastify.close((err) => {
+    t.error(err)
+  })
+})
+
+test('onRoute hook should pass correct route', t => {
+  t.plan(5)
+  const fastify = Fastify()
+  fastify.addHook('onRoute', (route) => {
+    t.strictEqual(route.method, 'GET')
+    t.strictEqual(route.url, '/')
+  })
+
+  fastify.register((instance, opts, next) => {
+    instance.addHook('onRoute', (route) => {
+      t.strictEqual(route.method, 'GET')
+      t.strictEqual(route.url, '/')
+    })
+    instance.get('/', opts, function (req, reply) {
+      reply.send()
+    })
+    next()
+  })
+
+  fastify.ready(err => {
+    t.error(err)
+  })
```

Comment:
The assertions don't make us sure that the both hooks are called. If the "external" hook is called twice the test passes.

Reply:
The first test for the route hook will guarantee that the hook is called only once.

Reply:
> The first test for the route hook will guarantee that the hook is called only once.

The first test tests that if only one hook is register the handler is called once.

For example:
```
for (var i = 0; i < onRouteHandlers.lenght; i++) {
  // notice the [0] instead of [i]
  onRouteHandlers[0].call(fastify, route)
}
```
passes both tests.

Reply:
Done